pkg/policy/cell/policy_importer.go

```diff
@@ -0,0 +1,401 @@
+	updates := stream.ToChannel(ctx, stream.Buffer(
+		stream.FromChannel(i.q),
+		option.Config.PolicyQueueSize, 10*time.Millisecond,
+		concat))
+
+	for {
+		select {
+		case <-ctx.Done():
+			for range updates {
+			}
+			return
+		case batch := <-updates:
+			i.processUpdates(ctx, batch)
+		}
+	}
+}
+
+func concat(buf []*policytypes.PolicyUpdate, in *policytypes.PolicyUpdate) []*policytypes.PolicyUpdate {
+	buf = append(buf, in)
+	return buf
+}
+
+// updatePrefixes determines the set of prefixes "owned" by a given resource and applies them
+// to the ipcache.
+// Write lock must be held.
+//
+// If the ipcache has started, it waits up to a configurable deadline for the prefixes
+// to be allocated.
+//
+// It returns the set of stale prefixes that should be deallocated after policy updates are complete.
+func (i *policyImporter) updatePrefixes(ctx context.Context, updates []*policytypes.PolicyUpdate) (toPrune map[ipcachetypes.ResourceID][]netip.Prefix) {
+	if i.ipc == nil {
+		return
+	}
+
+	// The set of all prefixes that belong to a resource.
+	toAllocate := map[ipcachetypes.ResourceID][]netip.Prefix{}
+	prefixSource := map[ipcachetypes.ResourceID]source.Source{}
+	toPrune = map[ipcachetypes.ResourceID][]netip.Prefix{}
+
+	// First, gather all new prefixes and index them by resource.
+	//
+	// For rules without an owning resource (k8s policy or file on disk),
+	// we never de-allocate prefixes. Otherwise, we will track prefixes
+	// by resource and de-allocate unused prefixes after policy is applied.
+	for _, upd := range updates {
+		prefixes := policy.GetCIDRPrefixes(upd.Rules)
+		if upd.Resource == "" {
+			// edge-case: no owning resource.
+			// Allocate prefixes with a placeholder.
+			if len(prefixes) == 0 {
+				continue
+			}
+			// since anonymous prefixes may come from multiple sources,
+			// we append to the list
+			toAllocate[ResourceIDAnonymous] = append(toAllocate[ResourceIDAnonymous], prefixes...)
+
+		} else {
+			// Standard case: there is an owning resource.
+			// Track the complete set of per-prefix resources.
+			// We want empty sets here!
+			toAllocate[upd.Resource] = prefixes
+			prefixSource[upd.Resource] = upd.Source
+		}
+	}
+
+	// Now that we know the exact set of prefixes for each resource, determine ipcache update.
+	// Note that we did this step separately, as we could have batched multiple updates
+	// for the same resource (i.e A1, A2, A3 -> only A3 matters)
+	//
+	// We elide updates when the prefix already has an entry for the given resource.
+	var ipcUpdates []ipcache.MU
+	for resource, newPrefixes := range toAllocate {
+		// For anonymous prefixes, just allocate, don't bookkeep
+		if resource == ResourceIDAnonymous {
+			for _, prefix := range newPrefixes {
+				ipcUpdates = append(ipcUpdates, ipcache.MU{
+					Prefix:   prefix,
+					Source:   prefixSource[resource],
+					Resource: resource,
+					Metadata: []ipcache.IPMetadata{labels.GetCIDRLabels(prefix)},
+				})
+			}
+			continue
```

Comment:
Correct, because we have no resource to "hold" them. There's no way to know if multiple anonymous policies allocate the same CIDR.

Previously, if two api-provided policies allocate the same CIDR, then one is deleted, that CIDR will be deallocated :-(.

Reply:
> How worried are we about long-running processes leaking lots of IPCache entires?

I don't have a lot of data, but this would only affect policies created by the local REST api. I don't think that sees a lot of usage, maybe @oblazek has an idea, since he uses it more.

> I also wonder if we could fix this by having a refcount for each distinct labelset.

We could, though the bookkeeping quickly gets tricky if you delete policies by labels that were created by resource. Especially if the policies in a resource have disjoint labels. Basically, it's a mess that mostly doesn't matter since nobody uses this anyways. I thought about this for a while and decided leaking prefixes (which has no correctness implications) is easier :-). But I could be convinced.

Reply:
I think this could get solved if we made the "anonymous" policies...well no longer anonymous. We could name them based on the policy revision/timestamp and that way, we can actually track them. But as mentioned in this thread already, it hasn't yet proved to be a problem.